lib/elastomer/client/docs.rb

```diff
@@ -269,8 +269,40 @@ def count( query, params = nil )
       def delete_by_query( query, params = nil )
         query, params = extract_params(query) if params.nil?
 
-        response = client.delete '/{index}{/type}/_query', update_params(params, :body => query, :action => 'docs.delete_by_query')
-        response.body
+        responses = []
+        accumulate = lambda { |response| responses << response unless response == nil }
+
+        accumulate.call(bulk do |b|
+          scan(query, params).each_document do |hit|
+            accumulate.call(b.delete(_id: hit["_id"]))
+          end
+        end)
+
+        response_items = responses.flat_map { |r| r["items"].map { |i| i["delete"] } }
+
+        is_ok = lambda { |status| status.between?(200, 299) }
+
+        categorize = lambda do |items|
+          {
+            "found" => items.count { |i| i["found"] },
+            "deleted" => items.count { |i| is_ok.call(i["status"]) },
+            "missing" => items.count { |i| !i["found"] },
+            "failed" => items.count { |i| i["found"] && !is_ok.call(i["status"]) },
+          }
+        end
+
+        indices = response_items
+          .map { |i| { i["_index"] => [i.tap { |ip| ip.delete("_index") }] } }
+          .reduce({}) { |acc, i| acc.merge(i) { |_, old, new| old + new } }
+          .map { |index, items| [index, categorize.call(items)] }.to_h
```

Comment:
I'd add a comment above each line explaining what its purpose is. I'd also comment on why accumulate is called both inside and outside the bulk block, and what the purpose of the `flat_map` line is (since `flat_map` is rarely used)